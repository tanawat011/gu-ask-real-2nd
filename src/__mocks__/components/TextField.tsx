// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck

export const TextField = ({ dataTestid, ...more }) => {
  Object.entries(more).forEach(([key, value]) => {
    if (typeof more[key] === 'boolean') {
      if (key === 'disabled') {
        if (more[key]) more[key] = true
        else delete more[key]

        return
      }

      more[key] = `${value}`
    }
  })

  return <input data-testid={dataTestid || 'text-field'} {...more} />
}
