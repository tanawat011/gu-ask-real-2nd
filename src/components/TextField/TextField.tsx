import type {
  DefaultEffectProps,
  DefaultInputProps,
  DefaultProps,
  VariantInput,
  defaultInputEventProps,
} from 'types'

import { BsExclamationCircleFill } from 'react-icons/bs'

import { useGetThemeColor } from 'hooks/useGetThemeColor'

import {
  TwInput,
  TwLabel,
  TwWrapperError,
  TwWrapperIconError,
  TwWrapperInput,
  TwWrapperInputLv2,
} from './TextField.style'

export type TextFieldProps = DefaultProps<VariantInput> &
  DefaultEffectProps &
  DefaultInputProps &
  defaultInputEventProps

export const TextField: React.FC<TextFieldProps> = ({
  id,
  name,
  label,
  value,
  placeholder,
  variant = 'basic',
  color,
  forceColor = false,
  disabled,
  size = 'md',
  shape = 'rounded',
  width = '185px',
  block = false,
  required,
  error = false,
}) => {
  const { hexColor, themeMode } = useGetThemeColor(color)

  const isBasic = variant === 'basic'

  const RenderLabel = () => (
    <TwLabel htmlFor={id} inputSize={size}>
      <span>{label}</span>
      {required && <span>*</span>}
    </TwLabel>
  )

  return (
    <TwWrapperInput block={block} shape={shape} width={width}>
      {label && isBasic && <RenderLabel />}

      <TwWrapperInputLv2
        variant={variant}
        hexColor={hexColor}
        forceColor={forceColor}
        themeMode={themeMode}
        error={error}
      >
        <TwInput
          id={id}
          name={name}
          value={value}
          placeholder={isBasic ? placeholder : ' '}
          themeMode={themeMode}
          hexColor={hexColor}
          forceColor={forceColor}
          disabled={disabled}
          variant={variant}
          inputSize={size}
          shape={shape}
          width={width}
          block={block}
          error={error}
          label={label}
        />

        {label && !isBasic && <RenderLabel />}

        {(!label || (label && !isBasic)) && error && (
          <TwWrapperIconError
            themeMode={themeMode}
            hexColor={hexColor}
            variant={variant}
            data-testid='text-field-error-icon'
          >
            <BsExclamationCircleFill />
          </TwWrapperIconError>
        )}

        {error && typeof error === 'string' && (
          <>
            <TwWrapperError>{error}</TwWrapperError>
          </>
        )}
      </TwWrapperInputLv2>
    </TwWrapperInput>
  )
}
