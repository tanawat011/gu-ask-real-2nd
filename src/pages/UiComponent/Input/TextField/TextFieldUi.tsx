import { useUiDisplayLayoutContext } from 'hooks/UiDisplayLayout'

import { menuItems, containerId, buttonApiTableData } from './TextFieldUi.data'
import { TextFieldUiBlock } from './TextFieldUiBlock'
import { TextFieldUiColor } from './TextFieldUiColor'
import { TextFieldUiDisabled } from './TextFieldUiDisabled'
import { TextFieldUiError } from './TextFieldUiError'
import { TextFieldUiForceColor } from './TextFieldUiForceColor'
import { TextFieldUiLabel } from './TextFieldUiLabel'
import { TextFieldUiRequired } from './TextFieldUiRequired'
import { TextFieldUiShape } from './TextFieldUiShape'
import { TextFieldUiSize } from './TextFieldUiSize'
import { TextFieldUiVariant } from './TextFieldUiVariant'
import { TextFieldUiWidth } from './TextFieldUiWidth'

export const TextFieldUi = () => {
  useUiDisplayLayoutContext({
    title: 'Text Field',
    description: 'Text field is used to get user input',
    containerId,
    apiId: menuItems.api.id,
    apiName: menuItems.api.name,
    apiData: buttonApiTableData as never,
    menuItems: menuItems,
  })

  return (
    <>
      <TextFieldUiVariant {...menuItems.variant} />
      <TextFieldUiSize {...menuItems.size} />
      <TextFieldUiShape {...menuItems.shape} />
      <TextFieldUiLabel {...menuItems.label} />
      <TextFieldUiError {...menuItems.error} />
      <TextFieldUiDisabled {...menuItems.disabled} />
      <TextFieldUiRequired {...menuItems.required} />
      <TextFieldUiWidth {...menuItems.width} />
      <TextFieldUiBlock {...menuItems.block} />
      <TextFieldUiColor {...menuItems.color} />
      <TextFieldUiForceColor {...menuItems.forceColor} />
    </>
  )
}
