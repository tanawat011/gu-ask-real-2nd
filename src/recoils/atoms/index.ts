export * from './pageUiComponents'
export * from './localSetting'
export * from './sidebar'
