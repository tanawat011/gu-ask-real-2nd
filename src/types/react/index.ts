export * from './children'
export * from './reactRouterDom'
