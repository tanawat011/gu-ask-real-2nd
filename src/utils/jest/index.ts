export * from './renderWithProviders'
export * from './twColor'
