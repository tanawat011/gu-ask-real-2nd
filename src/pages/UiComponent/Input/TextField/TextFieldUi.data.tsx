import type { ButtonProps } from 'components/Common'
import type { ComponentApiGuideData } from 'types'

import { PreCode } from 'components/PreCode'

export const containerId = 'button-ui'

export const menuItems = {
  variant: { id: 'variant', name: 'Variant' },
  size: { id: 'size', name: 'Size' },
  shape: { id: 'shape', name: 'Shape' },
  label: { id: 'label', name: 'Label' },
  error: { id: 'error', name: 'Error' },
  disabled: { id: 'disabled', name: 'Disabled' },
  required: { id: 'required', name: 'Required' },
  width: { id: 'width', name: 'Width' },
  block: { id: 'block', name: 'Block' },
  color: { id: 'color', name: 'Color' },
  forceColor: { id: 'force-color', name: 'Force Color' },
  helperText: { id: 'helperText', name: 'HelperText' }, // * string, ReactNode
  type: { id: 'type', name: 'Type' }, // * email, password, number, tel, url, search
  multiline: { id: 'multiline', name: 'Multiline' }, // * textarea
  margin: { id: 'margin', name: 'Margin' }, // * none, dense, normal
  icon: { id: 'icon', name: 'Icon' }, // * string, ReactNode
  iconPosition: { id: 'iconPosition', name: 'IconPosition' }, // * left, right
  loading: { id: 'loading', name: 'Loading' }, // * boolean, string
  api: { id: 'api', name: 'API' },
}

export const buttonApiTableData: ComponentApiGuideData<ButtonProps>[] = [
  {
    prop: 'label',
    desc: 'Text label of the button',
    val: <PreCode text='string' />,
    default: '-',
  },
  {
    prop: 'icon',
    desc: 'Render icon',
    val: (
      <>
        <PreCode text='string' />, <PreCode text='ReactNode' />
      </>
    ),
    default: '-',
  },
  {
    prop: 'iconPosition',
    desc: 'Icon prop can set position',
    val: (
      <>
        <PreCode text='left' /> or <PreCode text='right' />
      </>
    ),
    default: <PreCode text='left' />,
  },
  {
    prop: 'onClick',
    desc: 'Handler event when button clicked',
    val: (
      <>
        <PreCode text='(event) => void' />
      </>
    ),
    default: '-',
  },
  {
    prop: 'variant',
    desc: 'The variant of the button',
    val: (
      <>
        <PreCode text='primary' />, <PreCode text='secondary' />,{' '}
        <PreCode text='tertiary' />, <PreCode text='outline' />,{' '}
        <PreCode text='shadow' />, <PreCode text='text' />,{' '}
        <PreCode text='link' />
      </>
    ),
    default: <PreCode text='primary' />,
  },
  {
    prop: 'color',
    desc: 'Custom color of the button',
    val: (
      <>
        available colors option based on tailwind{' '}
        <PreCode text='theme.colors' />
      </>
    ),
    default: '-',
  },
  {
    prop: 'size',
    desc: 'The size of the button',
    val: (
      <>
        <PreCode text='xs' />, <PreCode text='sm' />, <PreCode text='md' />,{' '}
        <PreCode text='lg' />
      </>
    ),
    default: <PreCode text='md' />,
  },
  {
    prop: 'shape',
    desc: 'The border radius of the button',
    val: (
      <>
        <PreCode text='square' />, <PreCode text='rounded' />,{' '}
        <PreCode text='circle' />
      </>
    ),
    default: <PreCode text='rounded' />,
  },
  {
    prop: 'disabled',
    desc: 'Prevents user interact with the button',
    val: <PreCode text='boolean' />,
    default: <PreCode text='false' />,
  },
  {
    prop: 'loading',
    desc: 'Prevents user interact with the button and show loading indicator',
    val: (
      <>
        <PreCode text='boolean' /> or <PreCode text='string' />
      </>
    ),
    default: <PreCode text='false' />,
  },
  {
    prop: 'block',
    desc: 'Make the button full width',
    val: <PreCode text='boolean' />,
    default: <PreCode text='false' />,
  },
  {
    prop: 'width',
    desc: 'Make the button fixed width',
    val: <PreCode text='string' />,
    default: <PreCode text='-' />,
  },
]
