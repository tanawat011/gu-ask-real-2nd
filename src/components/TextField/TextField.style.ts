import type { TextFieldProps } from './TextField'
import type {
  Size,
  ThemeMode,
  TwShapeObject,
  TwSizeObject,
  WithRequired,
} from 'types'

import colors from 'tailwindcss/colors'
import tw, { css, styled } from 'twin.macro'

import { twColor } from 'utils/jest'

type TwThemeOption = {
  hexColor: string
  themeMode: ThemeMode
}

type TwWrapperInputProps = Required<
  Pick<TextFieldProps, 'block' | 'shape' | 'width'>
>

type TwWrapperInput2Props = Required<
  Pick<TextFieldProps, 'variant' | 'error'>
> &
  Pick<TextFieldProps, 'forceColor'> &
  TwThemeOption

type GetColorsFn = (option: TwWrapperInput2Props) => {
  bg: string
  bgInput: string
  bgInputDisabled: string
  border: string
  borderTheme: string
  text: string
  textTheme: string
}

type TwTextFieldProps = WithRequired<
  Omit<TextFieldProps, 'disabled' | 'placeholder' | 'size'>,
  'shape' | 'width' | 'error' | 'variant'
> &
  TwThemeOption & {
    inputSize: Size
  }

type TwLabelProps = {
  inputSize: Size
}

type TwWrapperIconErrorProps = Required<Pick<TextFieldProps, 'variant'>> &
  TwThemeOption

const twSize: TwSizeObject = {
  xs: tw`h-7 py-1 text-xs`,
  sm: tw`h-9 py-2 text-sm`,
  md: tw`h-11 py-2 text-sm`,
  lg: tw`h-14 py-2 text-base`,
}

const twShape: TwShapeObject = {
  square: tw`rounded-none`,
  rounded: tw`rounded-md`,
  circle: tw`rounded-full`,
}

const twLabelSize: TwSizeObject = {
  xs: tw`text-xs`,
  sm: tw`text-sm`,
  md: tw`text-base`,
  lg: tw`text-lg`,
}

const getColors: GetColorsFn = ({ themeMode, hexColor, error, variant }) => {
  const white = colors.white
  const darkBlack = colors.gray[900]
  const black = colors.gray[800]
  const darkedGray = colors.gray[700]
  const darkGray = colors.gray[600]
  const gray = colors.gray[500]
  const lightGray = colors.gray[300]
  const whiteGray = colors.gray[100]
  const red = colors.red[600]
  const base = hexColor

  const isDarkMode = themeMode === 'dark'
  const isFilled = variant === 'filled'

  const bgInputDisabled = twColor(isDarkMode ? darkedGray : lightGray)

  const bgInputDarkMode = isFilled ? darkBlack : colors.transparent
  const bgInputLightMode = isFilled ? whiteGray : colors.transparent
  const bgInput = twColor(isDarkMode ? bgInputDarkMode : bgInputLightMode)
  const bg = twColor(isDarkMode ? black : white)

  const borderError = twColor(red, 'border')
  const borderTheme = error ? borderError : twColor(base, 'border')
  const border = error
    ? borderError
    : twColor(isDarkMode ? darkGray : lightGray, 'border')

  const textError = twColor(red, 'text')
  const textTheme = error ? textError : twColor(base, 'text')
  const text = error ? textError : twColor(isDarkMode ? white : gray, 'text')

  return {
    bg,
    bgInput,
    bgInputDisabled,
    borderTheme,
    border,
    text,
    textTheme,
  }
}

export const TwWrapperInput = styled.div<TwWrapperInputProps>(
  ({ block, shape, width }) => {
    return [
      block
        ? tw`w-full`
        : css`
            width: ${width};
          `,
      tw`mb-7`,
      twShape[shape],
    ]
  },
)

export const TwWrapperInputLv2 = styled.div<TwWrapperInput2Props>(
  ({ variant, hexColor, forceColor, themeMode, error }) => {
    const { bg, text, textTheme } = getColors({
      themeMode,
      hexColor,
      forceColor,
      error,
      variant,
    })
    const isBasic = variant === 'basic'
    const isOutline = variant === 'outline'
    const isUnderline = variant === 'underline'
    const isFilled = variant === 'filled'

    return [
      tw`relative`,
      !isBasic &&
        css`
          label {
            ${tw`pointer-events-none absolute top-1/2 z-10 m-0 -translate-y-1/2 transition-all`}
            ${isOutline && tw`left-3`}
            ${isUnderline && tw`left-0`}
            ${isFilled && tw`left-3`}
            ${error && `color: ${text};`}
          }

          input:focus + label {
            color: ${textTheme};
          }

          input:focus + label,
          input:not(:placeholder-shown) + label {
            ${tw`-top-0.5 left-2 -translate-y-1/2 text-xs`}
            ${isOutline && tw`left-2 px-1`}
            ${isUnderline && tw`left-0`}
            ${isFilled && tw`left-3 -top-2`}
            background-color: ${bg};
          }
        `,
    ]
  },
)

export const TwLabel = styled.label<TwLabelProps>(({ inputSize }) => [
  tw`mb-2 block`,
  twLabelSize[inputSize],
  css`
    span:nth-child(2) {
      ${tw`ml-1 text-red-600`}
    }
  `,
])

export const TwWrapperIconError = styled.div<TwWrapperIconErrorProps>(
  ({ variant }) => [
    variant === 'underline' ? tw`right-1` : tw`right-3`,
    tw`absolute top-1/2 -translate-y-1/2 text-red-600`,
  ],
)
export const TwWrapperError = tw.div`absolute text-red-600`

export const TwInput = styled.input<TwTextFieldProps>(
  ({
    variant,
    hexColor,
    forceColor,
    themeMode,
    disabled,
    inputSize,
    shape,
    width,
    block,
    error,
    label,
  }) => {
    const { border, borderTheme, text, bgInput, bgInputDisabled } = getColors({
      themeMode,
      hexColor,
      forceColor,
      error,
      variant,
    })
    const isBasic = variant === 'basic'
    const isUnderline = variant === 'underline'
    const isFilled = variant === 'filled'

    const handleSize = () => [
      !isUnderline && tw`px-3`,
      (isUnderline || isFilled) && [
        tw`border-b focus:focus-within:border-b-2`,
        twSize[inputSize],
      ],
      !isUnderline && !isFilled && [tw`border focus:ring-1`, twSize[inputSize]],
    ]

    const handleError = () => [
      error && isBasic && label && tw`pr-3`,
      error && isBasic && !label && tw`pr-9`,
      error && !isBasic && !isUnderline && tw`pr-9`,
      error && !isBasic && isUnderline && tw`pr-7`,
    ]

    return [
      tw`outline-none transition-all`,
      !isUnderline && twShape[shape],
      handleSize(),
      handleError(),
      css`
        ${block ? tw`w-full` : `width: ${width};`}

        background-color: ${bgInput};

        ${tw`border-opacity-100 text-opacity-100`}
        border-color: ${forceColor ? borderTheme : border};

        color: ${text};

        &:disabled {
          ${tw`cursor-not-allowed`}

          &::placeholder {
            ${tw`opacity-70`};
          }
        }
      `,
      css`
        &:focus:focus-within {
          border-color: ${borderTheme};

          ${tw`ring-opacity-100`}
          --tw-ring-color: ${borderTheme};
        }
      `,
      disabled &&
        css`
          background-color: ${bgInputDisabled};
        `,
    ]
  },
)
