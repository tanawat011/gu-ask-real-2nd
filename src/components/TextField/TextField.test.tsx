import type { Size, ThemeMode, VariantInput } from 'types'

import { useEffect } from 'react'

import { screen } from '@testing-library/react'
import { useRecoilState } from 'recoil'

import { localSettingAtom } from 'recoils/atoms'
import { renderWithProviders, twColor } from 'utils/jest'

import { TextField } from './TextField'

jest.mock('./TextField', () => jest.requireActual('./TextField'))

describe('<TextField />', () => {
  test('renders correctly and match snapshot', async () => {
    expect.assertions(1)

    const { asFragment } = renderWithProviders(<TextField />)

    expect(asFragment()).toMatchSnapshot()
  })

  test('should render with label', async () => {
    expect.assertions(1)

    renderWithProviders(<TextField label='Label' />)

    const label = screen.getByText(/Label/i)

    expect(label).toBeInTheDocument()
  })

  test('should render with field required option', async () => {
    expect.assertions(1)

    renderWithProviders(<TextField label='Label' required />)

    const label = screen.getByText(/\*/i)

    expect(label).toBeInTheDocument()
  })

  test('should render with placeholder', async () => {
    expect.assertions(1)

    renderWithProviders(<TextField label='Label' placeholder='placeholder' />)

    const input = screen.queryByPlaceholderText(/placeholder/i)

    expect(input).toBeInTheDocument()
  })

  test('should render with placeholder but not basic variant', async () => {
    expect.assertions(1)

    renderWithProviders(
      <TextField label='Label' placeholder='placeholder' variant='outline' />,
    )

    const input = screen.queryByPlaceholderText('')

    expect(input).toBeInTheDocument()
  })

  test('should render with error message', async () => {
    expect.assertions(3)

    renderWithProviders(<TextField label='Label' error='Please enter' />)

    const labelError = screen.getByText('Please enter')
    const input = screen.getByRole('textbox')

    expect(labelError).toBeInTheDocument()
    expect(input).toHaveStyleRule('padding-right', '0.75rem')
  })

  test('should render with error message but no label', async () => {
    expect.assertions(3)

    renderWithProviders(<TextField error='Please enter' />)

    const labelError = screen.getByText('Please enter')
    const input = screen.getByRole('textbox')

    expect(labelError).toBeInTheDocument()
    expect(input).toHaveStyleRule('padding-right', '2.25rem')
  })

  test('should render with error but not `basic` or `underline` variant', async () => {
    expect.assertions(3)

    renderWithProviders(
      <TextField label='Label' variant='outline' error='Please enter' />,
    )

    const iconError = screen.getByTestId('text-field-error-icon')
    const input = screen.getByRole('textbox')

    expect(iconError).toBeInTheDocument()
    expect(input).toHaveStyleRule('padding-right', '2.25rem')
  })

  test('should render with error with `underline` variant', async () => {
    expect.assertions(3)

    renderWithProviders(
      <TextField label='Label' variant='underline' error='Please enter' />,
    )

    const iconError = screen.getByTestId('text-field-error-icon')
    const input = screen.getByRole('textbox')

    expect(iconError).toBeInTheDocument()
    expect(input).toHaveStyleRule('padding-right', '1.75rem')
  })

  test('should render with block option', async () => {
    expect.assertions(1)

    renderWithProviders(<TextField label='text field' block />)

    const button = screen.getByRole('textbox')

    expect(button).toHaveStyle('width: 100%')
  })

  test('should render with force color', async () => {
    expect.assertions(2)

    renderWithProviders(
      <TextField label='text field' color='emerald-500' forceColor />,
    )

    const input = screen.getByRole('textbox')

    expect(input).toHaveStyleRule('border-color', twColor('#10B981', 'border'))
  })

  test('should render with no force color', async () => {
    expect.assertions(2)

    renderWithProviders(
      <TextField label='text field' color='emerald-500' forceColor={false} />,
    )

    const input = screen.getByRole('textbox')

    expect(input).toHaveStyleRule('border-color', twColor('#4B5563', 'border'))
  })

  test('should render with disabled input', async () => {
    expect.assertions(6)

    renderWithProviders(<TextField label='text field' disabled />)

    const input = screen.getByRole('textbox')

    expect(input).toHaveStyleRule('background-color', twColor('#374151'))
    expect(input).toHaveStyleRule('cursor', 'not-allowed', {
      modifier: ':disabled',
    })
    expect(input).toHaveStyleRule('opacity', '0.7', {
      modifier: ':disabled::placeholder',
    })
  })

  type SizeOption = {
    height: string
    paddingX: string
    paddingY: string
    fontSize: string
    lineHeight: string
  }

  type TestSize = { [key in Size]: SizeOption }

  const sizes: TestSize = {
    xs: {
      fontSize: '0.75rem',
      lineHeight: '1rem',
      height: '1.75rem',
      paddingX: '0.75rem',
      paddingY: '0.25rem',
    },
    sm: {
      fontSize: '0.875rem',
      lineHeight: '1.25rem',
      height: '2.25rem',
      paddingX: '0.75rem',
      paddingY: '0.5rem',
    },
    md: {
      fontSize: '0.875rem',
      lineHeight: '1.25rem',
      height: '2.75rem',
      paddingX: '0.75rem',
      paddingY: '0.5rem',
    },
    lg: {
      fontSize: '1rem',
      lineHeight: '1.5rem',
      height: '3.5rem',
      paddingX: '0.75rem',
      paddingY: '0.5rem',
    },
  }

  Object.entries(sizes).forEach(
    ([key, { fontSize, lineHeight, height, paddingX, paddingY }]) => {
      test(`renders correctly and size \`${key}\``, () => {
        expect.assertions(7)

        renderWithProviders(
          <TextField label='text field size' size={key as Size} />,
        )

        const button = screen.getByRole('textbox')

        expect(button).toHaveStyle(`height: ${height}`)
        expect(button).toHaveStyle(`padding-top: ${paddingY}`)
        expect(button).toHaveStyle(`padding-bottom: ${paddingY}`)
        expect(button).toHaveStyle(`padding-left: ${paddingX}`)
        expect(button).toHaveStyle(`padding-right: ${paddingX}`)
        expect(button).toHaveStyle(`font-size: ${fontSize}`)
        expect(button).toHaveStyle(`line-height: ${lineHeight}`)
      })
    },
  )

  type ColorTestOption = {
    bgColor?: string
    bgOpacity?: string
    borderWidth?: string
    borderBottomWidth?: string
    borderColor?: string
    borderColorFocus?: string
    borderOpacity?: string
    ringColorFocus?: string
    ringOpacityFocus?: string
    textColor?: string
    textOpacity?: string
  }

  const execTestCase = (
    key: VariantInput,
    {
      bgColor,
      bgOpacity,
      borderWidth,
      borderBottomWidth,
      borderColor,
      borderColorFocus,
      borderOpacity,
      ringColorFocus,
      ringOpacityFocus,
      textColor,
      textOpacity,
    }: ColorTestOption,
    mode: ThemeMode = 'dark',
  ) => {
    const Component = () => {
      const [{ auth, theme, sidebar }, setTheme] =
        useRecoilState(localSettingAtom)

      useEffect(() => {
        setTheme({ auth, sidebar, theme: { ...theme, mode } })
      }, [])

      return <TextField label='Text Field' variant={key as VariantInput} />
    }

    renderWithProviders(<Component />)

    const input = screen.getByRole('textbox')

    if (bgColor) {
      expect(input).toHaveStyleRule('background-color', twColor(bgColor))
    }

    if (bgOpacity) {
      expect(input).toHaveStyleRule('--tw-bg-opacity', bgOpacity)
    }

    if (borderWidth) {
      expect(input).toHaveStyleRule('border-width', borderWidth)
    }

    if (borderBottomWidth) {
      expect(input).toHaveStyleRule('border-bottom-width', borderBottomWidth)
    }

    if (borderColor) {
      expect(input).toHaveStyleRule(
        'border-color',
        twColor(borderColor, 'border'),
      )
    }

    if (borderColorFocus) {
      expect(input).toHaveStyleRule(
        'border-color',
        twColor(borderColorFocus, 'border'),
        { modifier: ':focus:focus-within' },
      )
    }

    if (borderOpacity) {
      expect(input).toHaveStyleRule('--tw-border-opacity', borderOpacity)
    }

    if (ringColorFocus) {
      expect(input).toHaveStyleRule(
        '--tw-ring-color',
        twColor(ringColorFocus, 'border'),
        {
          modifier: ':focus:focus-within',
        },
      )
    }

    if (ringOpacityFocus) {
      expect(input).toHaveStyleRule('--tw-ring-opacity', ringOpacityFocus, {
        modifier: ':focus:focus-within',
      })
    }

    if (textColor) {
      expect(input).toHaveStyleRule('color', twColor(textColor, 'text'))
    }

    if (textOpacity) {
      expect(input).toHaveStyleRule('--tw-text-opacity', textOpacity)
    }
  }

  type TestVariant = { [key in VariantInput]?: ColorTestOption }

  const variantsDark: TestVariant = {
    basic: {
      bgColor: 'transparent',
      borderColor: '#4B5563',
      borderColorFocus: '#4f46e5',
      borderOpacity: '1',
      borderWidth: '1px',
      ringColorFocus: '#4f46e5',
      ringOpacityFocus: '1',
      textColor: '#ffffff',
    },
    outline: {
      bgColor: 'transparent',
      borderColor: '#4B5563',
      borderColorFocus: '#4f46e5',
      borderOpacity: '1',
      borderWidth: '1px',
      ringColorFocus: '#4f46e5',
      ringOpacityFocus: '1',
      textColor: '#ffffff',
    },
    underline: {
      bgColor: 'transparent',
      borderColor: '#4B5563',
      borderColorFocus: '#4f46e5',
      borderOpacity: '1',
      borderBottomWidth: '1px',
      ringColorFocus: '#4f46e5',
      ringOpacityFocus: '1',
      textColor: '#ffffff',
    },
    filled: {
      bgColor: '#111827',
      borderColor: '#4B5563',
      borderColorFocus: '#4f46e5',
      borderOpacity: '1',
      borderBottomWidth: '1px',
      ringColorFocus: '#4f46e5',
      ringOpacityFocus: '1',
      textColor: '#ffffff',
    },
  }

  Object.entries(variantsDark).forEach(([key, val]) => {
    test(`renders correctly with dark theme with variant \`${key}\``, () => {
      expect.hasAssertions()

      execTestCase(key as VariantInput, val)
    })
  })

  const variantsLight: TestVariant = {
    basic: {
      bgColor: 'transparent',
      borderColor: '#D1D5DB',
      borderColorFocus: '#4f46e5',
      borderOpacity: '1',
      borderWidth: '1px',
      ringColorFocus: '#4f46e5',
      ringOpacityFocus: '1',
      textColor: '#6b7280',
    },
    outline: {
      bgColor: 'transparent',
      borderColor: '#D1D5DB',
      borderColorFocus: '#4f46e5',
      borderOpacity: '1',
      borderWidth: '1px',
      ringColorFocus: '#4f46e5',
      ringOpacityFocus: '1',
      textColor: '#6b7280',
    },
    underline: {
      bgColor: 'transparent',
      borderColor: '#D1D5DB',
      borderColorFocus: '#4f46e5',
      borderOpacity: '1',
      borderBottomWidth: '1px',
      ringColorFocus: '#4f46e5',
      ringOpacityFocus: '1',
      textColor: '#6b7280',
    },
    filled: {
      bgColor: '#F3F4F6',
      borderColor: '#D1D5DB',
      borderColorFocus: '#4f46e5',
      borderOpacity: '1',
      borderBottomWidth: '1px',
      ringColorFocus: '#4f46e5',
      ringOpacityFocus: '1',
      textColor: '#6b7280',
    },
  }

  Object.entries(variantsLight).forEach(([key, val]) => {
    test(`renders correctly with light theme with variant \`${key}\``, () => {
      expect.hasAssertions()

      execTestCase(key as VariantInput, val, 'light')
    })
  })
})
