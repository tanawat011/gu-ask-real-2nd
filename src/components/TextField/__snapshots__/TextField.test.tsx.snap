// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders correctly and match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  width: 185px;
  margin-bottom: 1.75rem;
  border-radius: 0.375rem;
}

.c1 {
  position: relative;
}

.c2 {
  outline: 2px solid transparent;
  outline-offset: 2px;
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  border-radius: 0.375rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  border-width: 1px;
  height: 2.75rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
  width: 185px;
  background-color: transparent;
  --tw-border-opacity: 1;
  --tw-text-opacity: 1;
  border-color: rgb(75 85 99 / var(--tw-border-opacity));
  color: rgb(255 255 255 / var(--tw-text-opacity));
}

.c2:focus {
  --tw-ring-offset-shadow: var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);
  --tw-ring-shadow: var(--tw-ring-inset) 0 0 0 calc(1px + var(--tw-ring-offset-width)) var(--tw-ring-color);
  box-shadow: var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow,0 0 #0000);
}

.c2:disabled {
  cursor: not-allowed;
}

.c2:disabled::-webkit-input-placeholder {
  opacity: 0.7;
}

.c2:disabled::-moz-placeholder {
  opacity: 0.7;
}

.c2:disabled:-ms-input-placeholder {
  opacity: 0.7;
}

.c2:disabled::placeholder {
  opacity: 0.7;
}

.c2:focus:focus-within {
  border-color: rgb(79 70 229 / var(--tw-border-opacity));
  --tw-ring-opacity: 1;
  --tw-ring-color: rgb(79 70 229 / var(--tw-border-opacity));
}

<div
    class="c0"
    shape="rounded"
    width="185px"
  >
    <div
      class="c1"
    >
      <input
        class="c2"
        shape="rounded"
        value=""
        width="185px"
      />
    </div>
  </div>
</DocumentFragment>
`;
